### DISCUSSION TITLE
"Please restart Ghidra for extension changes to take effect."

### DISCUSSION THREAD

**Discussion by astrelsky** [NationalSecurityAgency/ghidra#3068] [General]
Why? This isn't actually necessary is it? Please correct me if I'm wrong but I feel like Ghidra has everything necessary in place already to avoid having to do this.

**Comment by jpleasu** [comment]
You're almost correct.

TLDR... if an extension wants to implement an extension point, it A) needs to get on the classpath and B) needs to get instantiated by its "owning" plugin.  (A) only happens at startup and (B) often happens once.

Very early in startup (you can see the `main` method on the same page) the `lib` dir of each installed extension is added to the classpath [here](https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Framework/Utility/src/main/java/ghidra/GhidraLauncher.java#L59-L66).

During a GUI run (for example), the `GhidraLaunchable` is `ghidra.GhidraRun`, whose `launch` method ultimately calls `ClassSearcher.search`, which finds and _caches_ "extension point" classes on the classpath in `ClassSearcher#extensionPoints`.  Note: this happens once!

(an extension point is a class whose name ends with one of the suffixes defined in an `ExtensionPoint.manifest` file, like [this one](https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Features/Base/data/ExtensionPoint.manifest))

`ClassSearcher#getInstances` is then used throughout Ghidra, typically at plugin start, to instantiate relevant extension points.  Note: many plugins only do this once.

OSGi was introduced to Ghidra to handle the complexities of a "dynamic classpath".  It's currently only available through scripting.   Somtimes, you can replace an extension's support libs with a jar "support" bundle.  Creating a bundle from a bunch of jars or class files can seem complicated, but there are tools to help.


For a general "dynamically loadable extension points" we'd have to address A and B.

A straightforward path forward for (A) would be to make `BundleHost` a framework level service.

For (B),  users of `ClassSearcher` need updating.  This comes down to tolerating a more general lifecyle by extension points where there currently is only a very simple one (every extension point is available at application launch and survives until application close). 

Some extension points could be made entirely dynamic - e.g. when instances are loaded, used, and disposed, like analyzers.  Other extension points could be adapted to a "born after launch, never die" policy, like fields in the GUI.

If extension points declared their lifecycle, migrating could be done incrementally - the default lifecycle would be the current one.  Then we could update one an extension point a time.

Sooo.. almost.

**Comment by dragonmacher** [comment]
Oversimplifying the idea that @jpleasu has listed, having a dynamic model makes the code base a bit more complicated to work in, as it adds the requirement to understand the idea of a dynamic lifecycle.    

Prior to extensions, we used to have some support for refreshing dynamic content while running.    This older feature was not used much and added extra complexity to the system.   There still may be some individual clients that choose to update internal content dynamically by calling the `ClassSearcher`.

In the end, to avoid adding a layer of dependency management,  we felt it was easier to restart the tool when new plugins or extensions are added.

**Comment by astrelsky** [comment]
I thought that plugins were initialized when a tool starts. This actually might be the case but there is no differentiation between an extension point and a plugin in this context (as far as loading/unloading the classes) so it does makes sense to restart here.

A rather silly idea maybe would be to have an option to have Ghidra restart itself. This may save someone the 5 seconds of having to manually close and relaunch it 🤣

**Reply by jpleasu** [reply]
I'm not clear what you mean in your first statement, but the second I actually support 110%.. a restart button would let me restart without moving my hand from the mouse 👍🏻

**Reply by astrelsky** [reply]
> I'm not clear what you mean in your first statement, but the second I actually support 110%.. a restart button would let me restart without moving my hand from the mouse 👍🏻 
> 

Regarding the first statement, I basically forgot there is more to handling extensions than the `Plugin` portion. Since the plugins can be enabled/disabled in the code browser I saw no reason why they couldn't be installed/uninstalled without restarting. I wasn't looking at the full picture.

**Reply by jpleasu** [reply]
Gotcha.  Well, I'm glad you're thinking about and discussing this stuff :-)
